First attempt

jhdalek55 opened the pull request:
This is the first carry over Deployment copy from the Google doc. It includes text for the following:
Exceptional Operations: all
Key Management: all except for "Repository Keys" and introductory text.
Normal Operating Procedures: I brought over copy for "Updating Metadata and Images" and "Backup and Garbage Collection." 
Customizing Uptane: all
Additional Security: al

Hope I did this Pull Request right.

tkfu commented:
I've cleaned up the branch so this PR only has your commits in it, but I haven't had a chance to look at the content yet. I'll try to get it merged later today.

jhdalek55 commented:
That makes sense. I was a bit unsure of how to indicate a link or
cross-reference copy. I deleted a few references, but left a few to show
where a cross-reference was needed. Anyway, I'll amend this copy the next
time I work on this copy (which will probably be the beginning of next
week).

I do think we need a few volunteers to review the existing figures in the
document and determine if they are still valid as is, or if they need to be
re-drawn. I'd rather not drag over figures that we will just have to delete
or drastically modify later.  I was planning to ask for these volunteers at
our meeting next Wednesday.

I would appreciate hearing back from you (and a handful of others) about
the email I sent out Monday (4/1) about the best way to link issues to the
in-process Deployment copy. As we close out issues on the Standards, I
would like to begin to open relevant issues for the Deployment materials.
So, I think we need to secure a home for these issues linked to where the
copy will be reviewed. Rather than rewrite the email here, I'll ask you to
just read it when you get a chance. It was addressed to about five people,
all of whom have been active in reviewing and commenting on issues. If you
did not receive that email, let me know and I'll re-send.

Thanks, Jon.

On Thu, Apr 4, 2019 at 1:21 PM Jon Oster <notifications@github.com> wrote:

> *@tkfu* requested changes on this pull request.
>
> Almost all of my comments here are related to references to sections or
> pages that no longer exist. For each of those, we should either mark it
> with a TODO (so that it doesn't get merged into master and deployed by
> accident), or link to the correct section of the standard, or of the new
> deployment considerations. For images / diagrams, they can be extracted
> from the google doc and added here.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272267805&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=Feuo_he4vNeaRkDEYpIU4mmEKrvm2lNS-CXRBMFxRgg&e=>
> :
>
> >  ## Scope of an update
>
>  (bootloader, system images, applications, data, etc.)
>
>  ## Delta update strategies
>
> +In order to save bandwidth costs, Uptane allows an OEM to deliver updates as
> +delta  images. A delta image update contains only the code and / or data that differs from the previous image installed by the ECU. In order to use delta images, the OEM SHOULD make the following changes.
> +
> +The OEM SHOULD add two types of information to the custom targets metadata used by the director repository: (1) the algorithm used to apply a delta image, and (2) the targets metadata about the delta image. This is done so that ECUs know how to apply and verify the delta image. The director repository SHOULD also be modified to produce delta images, because Uptane does not require it to compute deltas by default. The director repository can use the vehicle version manifest and dependency resolution to determine the differences between the previous and latest images. If desired, then the director repository MAY encrypt the delta image.
> +
> +As these images are produced on demand by the director repository, primaries SHOULD download all delta and / or encrypted images only from that source. After full verification of metadata, primaries SHOULD also check whether delta images match the targets metadata from the director repository, just as they check whether encrypted images match the targets metadata from the director repository when using non-delta images.
> +
> +Finally, in order to install a delta image, an ECU SHOULD take one of the actions described in Table D.2a, depending on whether or not the delta image has been encrypted, and if the ECU has additional storage. Note that the OEM MAY use stream ciphers in order to enable on-the-fly decryption by ECUs that have no additional storage space. In this case, the ECU would decrypt the delta image as it is downloaded, then follow the remainder of the steps in the third box below.
>
> References to section numbers and tables that don't exist in this document.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272268019&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=A8OTwdvCJKEFPK2q7FJ1OZHElEHogqN65pIzDqrPbQ8&e=>
> :
>
> >  ## Scope of an update
>
>  (bootloader, system images, applications, data, etc.)
>
>  ## Delta update strategies
>
> +In order to save bandwidth costs, Uptane allows an OEM to deliver updates as
> +delta  images. A delta image update contains only the code and / or data that differs from the previous image installed by the ECU. In order to use delta images, the OEM SHOULD make the following changes.
> +
> +The OEM SHOULD add two types of information to the custom targets metadata used by the director repository: (1) the algorithm used to apply a delta image, and (2) the targets metadata about the delta image. This is done so that ECUs know how to apply and verify the delta image. The director repository SHOULD also be modified to produce delta images, because Uptane does not require it to compute deltas by default. The director repository can use the vehicle version manifest and dependency resolution to determine the differences between the previous and latest images. If desired, then the director repository MAY encrypt the delta image.
> +
> +As these images are produced on demand by the director repository, primaries SHOULD download all delta and / or encrypted images only from that source. After full verification of metadata, primaries SHOULD also check whether delta images match the targets metadata from the director repository, just as they check whether encrypted images match the targets metadata from the director repository when using non-delta images.
> +
> +Finally, in order to install a delta image, an ECU SHOULD take one of the actions described in Table D.2a, depending on whether or not the delta image has been encrypted, and if the ECU has additional storage. Note that the OEM MAY use stream ciphers in order to enable on-the-fly decryption by ECUs that have no additional storage space. In this case, the ECU would decrypt the delta image as it is downloaded, then follow the remainder of the steps in the third box below.
> +
> +### Dynamic delta updates vs. precomputed delta updates
> +
> +There are two options when computing delta updates. Delta updates can be computed dynamically for each ECU during the installation process (dynamic delta updates), or possible delta images can be precomputed before installation begins (precomputed delta updates).
> +
> +Dynamic delta updates reduce the amount of data sent on each update, while allowing for fine grained control of what version is placed on each ECU.  By using the custom field of the targets metadata, as described in section D.9, the director can be configured to specify a particular version of software for every ECU. Dynamic delta updates allow the director to do  file granular resource tracking, which can save bandwidth by only transmitting the delta of the image.
>
> Reference to a section that doesn't exist.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272268126&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=5T3SSlPI1bsHvABOdgFX76oXUWj2fHaEVlboJYcs4s8&e=>
> :
>
> > +delta  images. A delta image update contains only the code and / or data that differs from the previous image installed by the ECU. In order to use delta images, the OEM SHOULD make the following changes.
> +
> +The OEM SHOULD add two types of information to the custom targets metadata used by the director repository: (1) the algorithm used to apply a delta image, and (2) the targets metadata about the delta image. This is done so that ECUs know how to apply and verify the delta image. The director repository SHOULD also be modified to produce delta images, because Uptane does not require it to compute deltas by default. The director repository can use the vehicle version manifest and dependency resolution to determine the differences between the previous and latest images. If desired, then the director repository MAY encrypt the delta image.
> +
> +As these images are produced on demand by the director repository, primaries SHOULD download all delta and / or encrypted images only from that source. After full verification of metadata, primaries SHOULD also check whether delta images match the targets metadata from the director repository, just as they check whether encrypted images match the targets metadata from the director repository when using non-delta images.
> +
> +Finally, in order to install a delta image, an ECU SHOULD take one of the actions described in Table D.2a, depending on whether or not the delta image has been encrypted, and if the ECU has additional storage. Note that the OEM MAY use stream ciphers in order to enable on-the-fly decryption by ECUs that have no additional storage space. In this case, the ECU would decrypt the delta image as it is downloaded, then follow the remainder of the steps in the third box below.
> +
> +### Dynamic delta updates vs. precomputed delta updates
> +
> +There are two options when computing delta updates. Delta updates can be computed dynamically for each ECU during the installation process (dynamic delta updates), or possible delta images can be precomputed before installation begins (precomputed delta updates).
> +
> +Dynamic delta updates reduce the amount of data sent on each update, while allowing for fine grained control of what version is placed on each ECU.  By using the custom field of the targets metadata, as described in section D.9, the director can be configured to specify a particular version of software for every ECU. Dynamic delta updates allow the director to do  file granular resource tracking, which can save bandwidth by only transmitting the delta of the image.
> +
> +To send dynamic delta updates, the director would compute the delta as described
> +earlier in this section. The computed images would be made available to the primary ECU using the non-standard direction process described in section D.9.
>
> Reference to a section that doesn't exist.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272268316&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=XGPB1EjOOO6U09FZY5XFVxVJyoXChls03ToKJD9cWV0&e=>
> :
>
> > +
> +As these images are produced on demand by the director repository, primaries SHOULD download all delta and / or encrypted images only from that source. After full verification of metadata, primaries SHOULD also check whether delta images match the targets metadata from the director repository, just as they check whether encrypted images match the targets metadata from the director repository when using non-delta images.
> +
> +Finally, in order to install a delta image, an ECU SHOULD take one of the actions described in Table D.2a, depending on whether or not the delta image has been encrypted, and if the ECU has additional storage. Note that the OEM MAY use stream ciphers in order to enable on-the-fly decryption by ECUs that have no additional storage space. In this case, the ECU would decrypt the delta image as it is downloaded, then follow the remainder of the steps in the third box below.
> +
> +### Dynamic delta updates vs. precomputed delta updates
> +
> +There are two options when computing delta updates. Delta updates can be computed dynamically for each ECU during the installation process (dynamic delta updates), or possible delta images can be precomputed before installation begins (precomputed delta updates).
> +
> +Dynamic delta updates reduce the amount of data sent on each update, while allowing for fine grained control of what version is placed on each ECU.  By using the custom field of the targets metadata, as described in section D.9, the director can be configured to specify a particular version of software for every ECU. Dynamic delta updates allow the director to do  file granular resource tracking, which can save bandwidth by only transmitting the delta of the image.
> +
> +To send dynamic delta updates, the director would compute the delta as described
> +earlier in this section. The computed images would be made available to the primary ECU using the non-standard direction process described in section D.9.
> +One drawback of dynamic delta updates is that if many ECUs are updating from the same version, computing the delta of each would result in duplicate computation that could be time consuming or take a lot of memory. A possible solution to this is to use precomputed delta updates.
> +
> +To send precomputed delta updates the director precomputes various probable diffs and makes these available as images. The director then specifies which precomputed delta image to send to each ECU by using the custom field of targets metadata, as described in section D.9. Precomputing the delta images has the added advantage of allowing these images to be stored on the image repository, which offers additional security against a director compromise.
>
> Reference to a section that doesn't exist.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272268956&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=AbwtpSLRIMZI7Zy98ILdP2B_po-Yc_Qvc6ykSxuLb5s&e=>
> :
>
> >  ## Uptane in conjunction with other protocols
> +Implementers MAY use Uptane in conjunction with other protocols already being used
> +to send updates to the vehicle.
> +
> +For example, implementers MAY use SSL / TLS to encrypt the connection between
> +primaries and the image and director repositories as well as the time server.
> +
> +For example, implementers MAY use OMA Device Management (OMA-DM) to send Uptane
> +metadata, images, and other messages to primaries.
> +
> +For example, implementers MAY Unified Diagnostic Services (UDS) to transport Uptane metadata, images, and other messages between primaries and secondaries.
> +
> +Any system being used to transport images to ECUs needs to modified only to also transport Uptane metadata, and other messages. Note that Uptane does not require network traffic between the director and image repositories and primaries, as well as between primaries and secondaries, to be authenticated.
> +
> +However, in order for an implementation to be Uptane-compliant, no ECU can cause another ECU to install an image without performing either full or partial verification of metadata. This is done in order to prevent attackers from being able to bypass Uptane, and thus execute arbitrary software attacks that way. Thus, in an Uptane-compliant implementation, an ECU performs either full or partial verification of metadata and images before installing any image (see Section 8 of the Implementation Specification), regardless of how the metadata and images were transmitted to the ECU.
>
> This doesn't use the keyword MUST, but it still has content that more
> properly belongs in the standard.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272269600&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=viNX-7DdKGJAEJdGbP_Oy2hSrzD12H-gTe3asnlWBEg&e=>
> :
>
> >  ## 2nd-party Fleet management
>
> +Some parties, such a vehicle rental company, or the military, may wish to exercise control on how their own fleet of vehicles are updated. We discuss two options for implementing fleet management using Uptane, illustrated by Figure D.6a. Both options use the map file (Section 2.5 of the Implementation Specification). Choosing between these options depends on whether the fleet manager wishes to have either complete control, or better compromise-resilience.
>
> Reference to a figure and section that doesn't exist.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272270059&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=r0iRFeVLM6l4cZBPFKGdcZ72qvXZ4NFYtOqoiROrIgI&e=>
> :
>
> >  ## User-customized updates
>
> +In its default implementation, Uptane allows only the OEM to fully control which updates are installed on which ECUs on which vehicles. Thus, no third party, such as a dealership, mechanic, fleet manager, or the end-user, has an input as to which updates should be installed. There are very good reasons, such as legal considerations, for enforcing this constraint. However, this capability exists to the point that the OEM wishes to make it available. We discuss two options for doing so.
> +
> +In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository would perform dependency resolution using this manifest, and propose a set of updates. In the third step, the third party would either agree with the OEM, or propose a different set of updates. This step SHOULD be authenticated (e.g., using client certificates, or username and password encrypted over TLS), so that only authorized third parties are allowed to negotiate with the OEM. In fourth step, the OEM would either agree with the third party, or propose a different set of updates. The third and fourth steps MAY be repeated, for a maximum number of retries, until both the OEM and the third party agree as to which updates should be installed.
>
> Reference to a figure that doesn't exist.
> ------------------------------
>
> In deployment/customizations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272270175&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=rYP-QAVO9uftvwrWLUzXBvVBBNCFTzx1d2vpRg6r4kY&e=>
> :
>
> >  ## User-customized updates
>
> +In its default implementation, Uptane allows only the OEM to fully control which updates are installed on which ECUs on which vehicles. Thus, no third party, such as a dealership, mechanic, fleet manager, or the end-user, has an input as to which updates should be installed. There are very good reasons, such as legal considerations, for enforcing this constraint. However, this capability exists to the point that the OEM wishes to make it available. We discuss two options for doing so.
> +
> +In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository would perform dependency resolution using this manifest, and propose a set of updates. In the third step, the third party would either agree with the OEM, or propose a different set of updates. This step SHOULD be authenticated (e.g., using client certificates, or username and password encrypted over TLS), so that only authorized third parties are allowed to negotiate with the OEM. In fourth step, the OEM would either agree with the third party, or propose a different set of updates. The third and fourth steps MAY be repeated, for a maximum number of retries, until both the OEM and the third party agree as to which updates should be installed.
> +
> +In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository would perform dependency resolution using this manifest, and propose a set of updates. In the third step, the third party would either agree with the OEM, or propose a different set of updates. This step SHOULD be authenticated (e.g., using client certificates, or username and password encrypted over TLS), so that only authorized third parties are allowed to negotiate with the OEM. In fourth step, the OEM would either agree with the third party, or propose a different set of updates. The third and fourth steps MAY be repeated, for a maximum number of retries, until both the OEM and the third party agree as to which updates should be installed.
>
> Reference to a figure that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272271446&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=zclapWnLLZ7lOlsN9ZkG0PeMq3WlDiFunmTCm59ONiA&e=>
> :
>
> >  ## Rolling back software
>
> +Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
> +
> +By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous targets metadata file. The first mechanism prevents an attacker from replaying an old metadata file. The second mechanism prevents an attacker who compromises the director repository from being able to choose old versions of images, despite being able to sign new metadata. See Figure C.1a for an example.
>
> Reference to a figure that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272271516&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=7N-vIrUr5pSSgoyEraZOVVYcS2gzoSp5Dq0_G8obF3M&e=>
> :
>
> >  ## Rolling back software
>
> +Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
> +
> +By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous targets metadata file. The first mechanism prevents an attacker from replaying an old metadata file. The second mechanism prevents an attacker who compromises the director repository from being able to choose old versions of images, despite being able to sign new metadata. See Figure C.1a for an example.
> +
> +There are at least two ways to solve this problem, each with different advantages and disadvantages.
> +
> +In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, an ECU would accept the installation of any compatible image referred in new targets metadata (see Section E.1).
>
> Reference to a figure and section that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272271587&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=KATgIg8Aa76IoyKD_gCX_SDs6jhq7D1eJxOWAypNt4E&e=>
> :
>
> >  ## Rolling back software
>
> +Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
> +
> +By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous targets metadata file. The first mechanism prevents an attacker from replaying an old metadata file. The second mechanism prevents an attacker who compromises the director repository from being able to choose old versions of images, despite being able to sign new metadata. See Figure C.1a for an example.
> +
> +There are at least two ways to solve this problem, each with different advantages and disadvantages.
> +
> +In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, an ECU would accept the installation of any compatible image referred in new targets metadata (see Section E.1).
> +
> +The advantage to this method is that it is simple. It allows the OEM to easily install interchangeable versions of the same image. In the example in Figure C.1b, “foo.img” may simply be a version of “bar.img” containing diagnostic functions. Therefore, the OEM may install either “bar.img” or “foo.img’ on the same ECU. The disadvantage to this method is that it allows attackers who compromise the director repository to execute rollback attacks, because obsolete images may also be installed. Therefore, this method SHOULD NOT be used.
>
> Reference to a figure that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272271618&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=fvcFheWyEQb-gZzS44skxvM1bZt0gs3pFmnU9UcunS8&e=>
> :
>
> >  ## Rolling back software
>
> +Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
> +
> +By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous targets metadata file. The first mechanism prevents an attacker from replaying an old metadata file. The second mechanism prevents an attacker who compromises the director repository from being able to choose old versions of images, despite being able to sign new metadata. See Figure C.1a for an example.
> +
> +There are at least two ways to solve this problem, each with different advantages and disadvantages.
> +
> +In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, an ECU would accept the installation of any compatible image referred in new targets metadata (see Section E.1).
> +
> +The advantage to this method is that it is simple. It allows the OEM to easily install interchangeable versions of the same image. In the example in Figure C.1b, “foo.img” may simply be a version of “bar.img” containing diagnostic functions. Therefore, the OEM may install either “bar.img” or “foo.img’ on the same ECU. The disadvantage to this method is that it allows attackers who compromise the director repository to execute rollback attacks, because obsolete images may also be installed. Therefore, this method SHOULD NOT be used.
> +
> +In the second option, an OEM MUST increment the release counter of an image whenever it is critical that an ECU not install images with lower release counters. In the example in Figure C.1c, if an ECU installs “foo.img”, then it cannot install “bar.img”. This is done to prevent the installation of compatible images with lower release counters that have known security vulnerabilities which have been fixed in newer images.
>
> Reference to a figure that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272271695&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=msW62yOgIkHhpzGdfTECv-PJYVXFu9rl0Hy1tI0mE1Y&e=>
> :
>
> >  ## Rolling back software
>
> +Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
> +
> +By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous targets metadata file. The first mechanism prevents an attacker from replaying an old metadata file. The second mechanism prevents an attacker who compromises the director repository from being able to choose old versions of images, despite being able to sign new metadata. See Figure C.1a for an example.
> +
> +There are at least two ways to solve this problem, each with different advantages and disadvantages.
> +
> +In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, an ECU would accept the installation of any compatible image referred in new targets metadata (see Section E.1).
> +
> +The advantage to this method is that it is simple. It allows the OEM to easily install interchangeable versions of the same image. In the example in Figure C.1b, “foo.img” may simply be a version of “bar.img” containing diagnostic functions. Therefore, the OEM may install either “bar.img” or “foo.img’ on the same ECU. The disadvantage to this method is that it allows attackers who compromise the director repository to execute rollback attacks, because obsolete images may also be installed. Therefore, this method SHOULD NOT be used.
> +
> +In the second option, an OEM MUST increment the release counter of an image whenever it is critical that an ECU not install images with lower release counters. In the example in Figure C.1c, if an ECU installs “foo.img”, then it cannot install “bar.img”. This is done to prevent the installation of compatible images with lower release counters that have known security vulnerabilities which have been fixed in newer images.
> +
> +The advantage to this method is that it prevents attackers who compromise only the director repository from being able to execute rollback attacks. However, there are two disadvantages. First, the release counters for images have to be maintained even if role B now signs for images previously signed by role A. This is because release counters are always compared to previous targets metadata files. Second, it is more cumbersome to roll back updates, or deliberately cause ECUs to install older images, because offline keys are used to increment the release counters of these older images in the new targets metadata for the image repository. However, this method SHOULD be preferred, because it is more secure. See Section E.2 for more techniques that can be used to limit rollback attacks when the director repository is compromised.
>
> Reference to a section that doesn't exist.
> ------------------------------
>
> In deployment/exceptional_operations.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272272026&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=f88Wqut93riZw_0cenUkqNFr2uryNIfOqR0dDzV_k2g&e=>
> :
>
> >  ## Adding or removing a supplier
>
> +Due to changes in business relationships, it may be the case that the OEM may need to add or remove a tier-1 supplier on its repositories.
> +
> +To add a tier-1 supplier, the OEM SHOULD use the following steps. First, if the supplier signs its own images, then the OEM SHALL add a delegation to the supplier on the image repository following the steps described in Section B.1. Second, the supplier SHALL deliver metadata and / or images to the OEM following the steps in Section A.1.1. Finally, the OEM SHALL add the metadata and images to its repositories, possibly test them, and then release them to affected vehicle following the steps in Section A.1.2.
>
> Reference to sections that don't exist.
> ------------------------------
>
> In deployment/key_management.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272272736&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=Ngo-yOEC0Gcrmh56C4f6d5DYDpw0JCkCSodGJqjA3ck&e=>
> :
>
> >
>  ### Key thresholds
>
> +Since the root role has the highest impact when its keys are compromised, it SHOULD use a sufficiently large threshold number of keys, so that a single key compromise is insufficient to sign its metadata file. Each key MAY belong to a repository administrator. For example, if there are 8 administrators, then at least 5 keys SHOULD be required to sign the root metadata file, so that a quorum is required to trust the metadata.
> +
> +The timestamp, snapshot, and targets roles MAY each use a single key, because using more keys does not add to security, since these keys are online, and attackers who compromise the repository can always use these online keys.
> +
>  ## ECU keys and the Inventory Database
>
> This whole section ("ECU keys and the Inventory Database") should actually
> be deleted--it's part of the standard.
> ------------------------------
>
> In deployment/normal_operation.md
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23discussion-5Fr272275174&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=B14-_21V2K925Gw25HAONSJAV8kJTHDZkyPOnAq-4M8&e=>
> :
>
> > +If the supplier signs its own images, then it delivers all of its metadata, including delegations, and associated images. Otherwise, if the OEM signs images on behalf of the supplier, then the supplier needs to update ONLY images, and the OEM is responsible for producing signed metadata. Regardless of whether it is the OEM or suppliers that produce signed metadata, the release counters associated with images SHOULD be incremented, so that attackers who compromise the director repository are not able to rollback to obsolete images (see the Additional Security page for more on this attack.)
> +
> +Regardless of the transport mechanism used to deliver images, the most important thing is for the OEM to somehow ensure that the images are authentic, and have not been tampered with by man-in-the-middle attackers. The OEM SHOULD verify these images using some out-of-band mechanism, so that the authenticity and integrity of these images can be double-checked.
> +For example, to obtain a higher degree of assurance, and for additional validation, the OEM MAY also require the supplier's update team to send a PGP / GPG signed email to the OEM's security team listing the cryptographic hashes of the new files.
> +
> +Alternatively, the OEM MAY require that updates be transmitted via DVD delivered by bonded and insured couriers. To validate the provided files, the OEM and a known contact at the supplier MAY have a video call where the cryptographic hashes of the metadata and / or images are provided by the supplier, and confirmed by the OEM.
> +
> +An OEM SHOULD perform this verification even if a trusted transport mechanism is used, because the OEM cannot be sure whether the transport mechanism has been compromised. If the suppliers have signed metadata, then the OEM SHOULD verify metadata and images by checking version numbers, expiration timestamps, delegations, signatures, and hashes, so that it can be sure that the metadata matches the images.
> +
> +### What the OEM should do
> +
> +After a tier-1 supplier has delivered new images to the OEM, and the OEM has somehow verified the authenticity and integrity of these images, the OEM SHOULD test new metadata and images before releasing them, in order to ensure that the images work as intended end user vehicles. To do so, It SHOULD use the following steps.
> +
> +First, the OEM SHOULD add these metadata and images to the image repository. It SHOULD also add information about these images to the inventory database, including any dependencies and conflicts between images for different ECUs. Both of these steps are done, to  make  the new metadata and images available to vehicles.
> +
> +Optionally, if images are encrypted on demand per ECU, then the OEM SHOULD ensure that the director repository somehow has access to the original, unencrypted images, so that automated processes running the director repository are able to encrypt them in the first place. It does not matter how the original, unencrypted images are stored on the director repository. For example, they MAY be stored unencrypted, or they MAY be encrypted using a master key that is known by the automated processes. See Section A.2.1 for more details.
>
> Reference to a section that does not exist.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-23pullrequestreview-2D222878112&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=MDk5Di5JCbm_n0TCe-rspN1bwoEFupks6JseKlQOj8Q&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AN5qX5dEVYbUKpqvoPeQQceb-5FqIhSw5Vks5vdjSTgaJpZM4cWuw0&d=DwMFaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=_V2IhS6Fwuv4lsKTh3_4bN81rUfvJgC9wHpfmQhPtKI&s=sVopkECnHUf1aLAXlsBmh5hMlHGEsyccQLEOk0l_GnE&e=>
> .
>

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+
+Any system being used to transport images to ECUs needs to modified only to also transport Uptane metadata, and other messages. Note that Uptane does not require network traffic between the director and image repositories and primaries, as well as between primaries and secondaries, to be authenticated.
+
+However, in order for an implementation to be Uptane-compliant, no ECU can cause another ECU to install an image without performing either full or partial verification of metadata. This is done in order to prevent attackers from being able to bypass Uptane, and thus execute arbitrary software attacks that way. Thus, in an Uptane-compliant implementation, an ECU performs either full or partial ver ...
```
So, what do I do with this? Should I move it to the standard and if so, where would it go? And, is it just the text highlighted here that needs to move?

jhdalek55 replied:
Still awaiting instructions on this. As I am prepping this copy, it would be very easy to open a pull request and insert it into the standard. But I would need to know exactly where it should go and if its just this text or the whole section.

jhdalek55 commented on deployment/key_management.md:
```diff
@@ -9,16 +9,61 @@ css_id: key_management
+
+The timestamp, snapshot, and targets roles MAY each use a single key, because using more keys does not add to security, since these keys are online, and attackers who compromise the repository can always use these online keys.
+
 ## ECU keys and the Inventory Database
```
Ok. When I go back in, I'll delete that.

jhdalek55 replied:
Section was deleted as requested.

jhdalek55 commented:
I will go back and acknowledge that each of the corrections above were fixed when I wrap everything up, but here is a quick status report. I have added two commits so far and will be adding two or three more probably tomorrow. I'm committing change to the Pull request, so hopefully this will be executable.

Offline, I have pulled out from the document all the figures except maybe three that may need edits and have saved them as JPEG files. I need to know whether I can add these graphics to the image file already on  the repo or if I have to open a new file on this branch. I don't believe I have ever set up a folder on a repo, but I'm sure I can figure it out.

I have put in TODOs for missing graphics and section links. I would much rather take care of importing graphics once the content has been merged and is not sitting in a pull request. I'd rather open a clean pull request to just take care of the graphics.

I have never made a table in Markdown. I was thinking of exporting the current tables into PDFs and then loading them like image files, but that would make it difficult if they have to be changed or updated (though I'm not sure if there would be a need to revise the tables here or just inset a new PDF.)

I've put in TODOs to link to sections as well. This has been a bit clumsy as, unlike the original, we don't have section numbers to refer to. I  know anchors can be put into place, or we can put in URL links for the specific parts of the web site where the text we want appears. But I don't see how to do the latter until the text is posted to the web site and therefore has an actual URL.

Other than trying to get rid of phantom section and image numbers, I've not done much editing except for fixing a few clumsy and repetitive sentences as I've imported. I think editing will definitely be needed but its probably better to do that after the content has been reviewed.

There are still empty sections, heading for which I could find no relevant content. I marked these with TODOs as well.

BTW, I hope I did the TODOs correctly, as I could find no reference as to how this should be done in Markdown. So I just followed an example I saw in the Standards file. 

I'm hoping to finish up what I can do (add all existing text, make sure all graphics and links are in place) later this week or at the latest early next week. I know that this has been slow, but I had a rather big learning curve to assimilate here.

Thanks.

Lois

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+
+As these images are produced on demand by the director repository, primaries SHOULD download all delta and / or encrypted images only from that source. After full verification of metadata, primaries SHOULD also check whether delta images match the targets metadata from the director repository, just as they check whether encrypted images match the targets metadata from the director repository when ...
+
+Finally, in order to install a delta image, an ECU SHOULD take one of the actions described in Table D.2a, depending on whether or not the delta image has been encrypted, and if the ECU has additional storage. Note that the OEM MAY use stream ciphers in order to enable on-the-fly decryption by ECUs that have no additional storage space. In this case, the ECU would decrypt the delta image as it is ...
```
I've added two commits to fix all of your comments. Two have been pushed to this PR. Another three or four commits will be added later today. Note that the commits still to be pushed to your branch will also have the remainder of the copy from the Google Doc.

Changed table # to Table 1  and inserted TODO to insert the existing table in that slot.

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+
+There are two options when computing delta updates. Delta updates can be computed dynamically for each ECU during the installation process (dynamic delta updates), or possible delta images can be precomputed before installation begins (precomputed delta updates).
+
+Dynamic delta updates reduce the amount of data sent on each update, while allowing for fine grained control of what version is placed on each ECU.  By using the custom field of the targets metadata, as described in section D.9, the director can be configured to specify a particular version of software for every ECU. Dynamic delta updates allow the director to do  file granular resource tracking, ...
```
Substituted a reference by name for the section number and posted a TODO to insert the appropriate link.

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+Dynamic delta updates reduce the amount of data sent on each update, while allowing for fine grained control of what version is placed on each ECU.  By using the custom field of the targets metadata, as described in section D.9, the director can be configured to specify a particular version of software for every ECU. Dynamic delta updates allow the director to do  file granular resource tracking, ...
+
+To send dynamic delta updates, the director would compute the delta as described
+earlier in this section. The computed images would be made available to the primary ECU using the non-standard direction process described in section D.9.
```
replaced with the phrase "described later in this section." As a link will be added to the same section in the paragraph above, I did not flag this as a link, but one could certainly be added later.

PLEASE NOTE, I did not attempt to set up anchors or some way to do direct linking in this version. But I did mark them as TODOs for later correction. Like the graphics, I think this is more easily done once the the copy is merged into the main branch.

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+earlier in this section. The computed images would be made available to the primary ECU using the non-standard direction process described in section D.9.
+One drawback of dynamic delta updates is that if many ECUs are updating from the same version, computing the delta of each would result in duplicate computation that could be time consuming or take a lot of memory. A possible solution to this is to use precomputed delta updates.
+
+To send precomputed delta updates the director precomputes various probable diffs and makes these available as images. The director then specifies which precomputed delta image to send to each ECU by using the custom field of targets metadata, as described in section D.9. Precomputing the delta images has the added advantage of allowing these images to be stored on the image repository, which off ...
```
Here again used phrase "later in the section"

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+
 ## 2nd-party Fleet management
 
+Some parties, such a vehicle rental company, or the military, may wish to exercise control on how their own fleet of vehicles are updated. We discuss two options for implementing fleet management using Uptane, illustrated by Figure D.6a. Both options use the map file (Section 2.5 of the Implementation Specification). Choosing between these options depends on whether the fleet manager wishes to ha ...
```
Labeled the figure as Figure 2 and left TODO note to insert the designated figure.

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
 
+In its default implementation, Uptane allows only the OEM to fully control which updates are installed on which ECUs on which vehicles. Thus, no third party, such as a dealership, mechanic, fleet manager, or the end-user, has an input as to which updates should be installed. There are very good reasons, such as legal considerations, for enforcing this constraint. However, this capability exists t ...
+
+In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository  ...
```
Labeled the figure as Figure 3 and left TODO to insert it.

jhdalek55 commented on deployment/customizations.md:
```diff
@@ -5,24 +5,117 @@ css_id: customizations
+
+In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository  ...
+
+In the first option, an OEM MAY elect to receive input from a third party as to which updates should be installed. A process for how this may be done is illustrated in Figure D.7a. In the first step, the vehicle would submit to the director repository controlled by the OEM its vehicle version manifest, which lists which updates are currently installed. In the second step, the director repository  ...
```
Also re-labeled Figure 3

jhdalek55 commented on deployment/exceptional_operations.md:
```diff
@@ -5,12 +5,44 @@ css_id: exceptional_operations
 
+Sometimes, it may be necessary to roll back updates. This is because, for example, if the latest updates are less reliable than previous ones, then the OEM may wish to rollback to the previous updates.
+
+By default, Uptane does not allow updates to be rolled back. There are two mechanisms that are used to achieve this. First, Uptane will reject any new metadata file with a version number lower than what is contained in the previous metadata file. Second, Uptane will reject any new image associated with a release counter that is lower than the release counter of the previous image in the previous  ...
```
Changed to Figure 1 and TODO added to insert the figure.

jhdalek55 commented on deployment/exceptional_operations.md:
```diff
@@ -5,12 +5,44 @@ css_id: exceptional_operations
+
+There are at least two ways to solve this problem, each with different advantages and disadvantages.
+
+In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, ...
```
Changed to Figure 2 and TODO added to insert the figure. Section number changed to "see the Security page on this website"

jhdalek55 commented on deployment/exceptional_operations.md:
```diff
@@ -5,12 +5,44 @@ css_id: exceptional_operations
+
+In the first option, an OEM MAY choose never to increment the release counters of images (see Figure C.1b). Uptane will accept any new image associated with a release counter that is equal to the release counter of the previous image in the previous targets metadata file. Therefore, if release counters are never incremented, then all images would have the same release counters. In this situation, ...
+
+The advantage to this method is that it is simple. It allows the OEM to easily install interchangeable versions of the same image. In the example in Figure C.1b, “foo.img” may simply be a version of “bar.img” containing diagnostic functions. Therefore, the OEM may install either “bar.img” or “foo.img’ on the same ECU. The disadvantage to this method is that it allows attackers who compromise the  ...
```
Changed to Figure 2 and TODO added to insert the figure. Section number changed to "see the Security page on this website"

jhdalek55 commented on deployment/exceptional_operations.md:
```diff
@@ -5,12 +5,44 @@ css_id: exceptional_operations
+
+The advantage to this method is that it is simple. It allows the OEM to easily install interchangeable versions of the same image. In the example in Figure C.1b, “foo.img” may simply be a version of “bar.img” containing diagnostic functions. Therefore, the OEM may install either “bar.img” or “foo.img’ on the same ECU. The disadvantage to this method is that it allows attackers who compromise the  ...
+
+In the second option, an OEM MUST increment the release counter of an image whenever it is critical that an ECU not install images with lower release counters. In the example in Figure C.1c, if an ECU installs “foo.img”, then it cannot install “bar.img”. This is done to prevent the installation of compatible images with lower release counters that have known security vulnerabilities which have be ...
```
Changed to Figure 2 in first paragraph and Figure 3  in the second and TODOs added to insert the figures.

jhdalek55 commented on deployment/exceptional_operations.md:
```diff
@@ -5,12 +5,44 @@ css_id: exceptional_operations
+
+In the second option, an OEM MUST increment the release counter of an image whenever it is critical that an ECU not install images with lower release counters. In the example in Figure C.1c, if an ECU installs “foo.img”, then it cannot install “bar.img”. This is done to prevent the installation of compatible images with lower release counters that have known security vulnerabilities which have be ...
+
+The advantage to this method is that it prevents attackers who compromise only the director repository from being able to execute rollback attacks. However, there are two disadvantages. First, the release counters for images have to be maintained even if role B now signs for images previously signed by role A. This is because release counters are always compared to previous targets metadata files ...
```
Changed section number to "See the Security pages." Left "TODO" to insert link to the relevant page.

jhdalek55 commented on deployment/normal_operation.md:
```diff
@@ -15,4 +15,40 @@ css_id: normal_operation
+
+First, the OEM SHOULD add these metadata and images to the image repository. It SHOULD also add information about these images to the inventory database, including any dependencies and conflicts between images for different ECUs. Both of these steps are done, to  make  the new metadata and images available to vehicles.
+
+Optionally, if images are encrypted on demand per ECU, then the OEM SHOULD ensure that the director repository somehow has access to the original, unencrypted images, so that automated processes running the director repository are able to encrypt them in the first place. It does not matter how the original, unencrypted images are stored on the director repository. For example, they MAY be stored  ...
```
Changed section number of name of the web page and included TODO to create the link.

jhdalek55 commented:
Note, newer commits are on the same pull request, but the branch name is pr/10.

tkfu commented:
I've moved all of this content over to the new repo at https://github.com/uptane/deployment-considerations/. I think we can close this now, and move all future PRs over to that repo.

jhdalek55 commented:
Thanks so much for getting this in place, Jon. Maybe in the next week or so
I can get the figures in place and then we can begin to work on the copy
for real,

Talk to you in a few minutes.

Lois

On Wed, Jul 17, 2019 at 10:31 AM Jon Oster <notifications@github.com> wrote:

> Closed #10
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=JJl46hLg4ID_7nzoeUQ5JWQz_H5b8woZIy17J3No1lg&s=1jWKiWeM7N8ltJk2UZDucp7MTrJU1ldQMBZAia2Uyms&e=>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_uptane_uptane.github.io_pull_10-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DADPGUX2KCZYIDXAKVQSEUVTP74ULVA5CNFSM4HC25Q2KYY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOSRSQIGQ-23event-2D2489648154&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=JJl46hLg4ID_7nzoeUQ5JWQz_H5b8woZIy17J3No1lg&s=PWwPWygrJCufet3pwhqUINdA1k--dRhkjffUZBDiyFE&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADPGUX446OMFMTQPMM4ECGTP74ULVANCNFSM4HC25Q2A&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=hgBKIqNYIOwzXeBjPUaKRw&m=JJl46hLg4ID_7nzoeUQ5JWQz_H5b8woZIy17J3No1lg&s=B9o44FaiLS15Df7nW-kr2ClkshPJANjKjB46j0eT3p4&e=>
> .
>
